src/gflownet/tasks/qm9_moo.py

```diff
@@ -0,0 +1,375 @@
+
+        scalarized_reward = self.pref_cond.transform(cond_info, flat_reward)
+        focused_reward = (
+            self.focus_cond.transform(cond_info, flat_reward, scalarized_reward)
+            if self.focus_cond is not None
+            else scalarized_reward
+        )
+        tempered_reward = self.temperature_conditional.transform(cond_info, focused_reward)
+        return RewardScalar(tempered_reward)
+
+    def compute_flat_rewards(self, mols: List[RDMol]) -> Tuple[FlatRewards, Tensor]:
+        graphs = [mxmnet.mol2graph(i) for i in mols]  # type: ignore[attr-defined]
+        assert len(graphs) == len(mols)
+        is_valid = torch.tensor([i is not None for i in graphs]).bool()
+        valid_graphs = [g for g in graphs if g is not None]
+        valid_mols = [m for m, g in zip(mols, graphs) if g is not None]
+        assert len(valid_mols) == len(valid_graphs)
+        if not is_valid.any():
+            return FlatRewards(torch.zeros((0, len(self.objectives)))), is_valid
+        else:
+            flat_r: List[Tensor] = []
+            for obj in self.objectives:
+                if obj == "gap":
+                    batch = gd.Batch.from_data_list(valid_graphs)
+                    batch.to(self.device)
+                    preds = self.models["mxmnet_gap"](batch)
+                    preds = preds.reshape((-1,)).data.cpu() / mxmnet.HAR2EV  # type: ignore[attr-defined]
+                    preds[preds.isnan()] = 1
+                elif obj == "qed":
+                    preds = torch.tensor([safe(QED.qed, i, 0) for i in valid_mols])
+                elif obj == "sa":
+                    preds = torch.tensor([safe(sascore.calculateScore, i, 10) for i in valid_mols])
+                elif obj == "mw":
+                    preds = torch.tensor([safe(Descriptors.MolWt, i, 1000) for i in valid_mols])
+                else:
+                    raise ValueError(f"{obj} not known")
+                flat_r.append(preds)
+                assert len(preds) == len(valid_graphs), f"{len(preds)} != {len(valid_graphs)} for obj {obj}"
+
+            flat_rewards = torch.stack(flat_r, dim=1)
+            return self.flat_reward_transform(flat_rewards), is_valid
+
+
+class QM9MOOTrainer(QM9GapTrainer):
+    task: QM9GapMOOTask
+    ctx: MolBuildingEnvContext
+
+    def set_default_hps(self, cfg: Config):
+        super().set_default_hps(cfg)
+        cfg.algo.sampling_tau = 0.95
+        # We use a fixed set of preferences as our "validation set", so we must disable the preference (cond_info)
+        # sampling and set the offline ratio to 1
+        cfg.algo.valid_sample_cond_info = False
+        cfg.algo.valid_offline_ratio = 1
+
+    def setup_task(self):
+        self.task = QM9GapMOOTask(
+            dataset=self.training_data,
+            cfg=self.cfg,
+            rng=self.rng,
+            wrap_model=self._wrap_for_mp,
+        )
+
+    def setup(self):
+        super().setup()
+        # self.sampling_hooks.append(
```

Comment:
Maybe add a flag or remove entirely

Reply:
For performance reasons, would it make sense to remove this in both moo classes?